### PR TITLE
[Model] Optimize Audex streaming decoder KV cache

### LINKED ISSUE #5964: [Performance]: Optimize Audex streaming decoder KV cache
### Proposal to improve performance

Audex's streaming causal speech decoder currently performs avoidable work for every decoded chunk:

- `CausalCodecDecoderCache.update` appends K/V with `torch.cat` in every transformer layer. Each append copies the full cache history, so cumulative cache-copy traffic grows quadratically with the generated sequence length.
- `RotaryPositionalEmbeddings.forward` calls `input_pos.max().item()` for both Q and K in every layer, introducing repeated device-to-host synchronization during streaming decode.
- The causal attention mask is identical across layers for a given chunk, but is currently rebuilt independently in each layer.

A proposed optimization is to:

1. Replace repeated concatenation with per-layer growable K/V buffers that append only the new chunk and expand geometrically when capacity is exhausted.
2. Prepare the shared RoPE cache once per decoder invocation using the Python-side cache position, avoiding tensor readback from the device.
3. Construct the causal mask once per chunk in the backbone and reuse it across transformer layers.

The cache should remain allocated per request/session rather than preallocating the maximum sequence length, so short or concurrent requests do not reserve worst-case KV memory.

### PR DESCRIPTION
**Pull request by pxljs**
## Purpose

Fixes #5964.

Audex performs streaming causal speech decoding in small chunks, but the existing cache path does repeated work in every transformer layer:

- K/V are appended with `torch.cat`, which copies the full history on every chunk and produces quadratic cumulative cache-copy traffic.
- RoPE reads `input_pos.max().item()` for Q and K in every layer, causing repeated device-to-host synchronization.
- The same causal mask is rebuilt independently in every layer.

This PR:

- replaces repeated concatenation with per-layer geometrically growing K/V buffers, copying only the new chunk on the steady-state path;
- prepares the shared RoPE cache once per decoder call from the Python-side cache position;
- builds the causal mask once per chunk and reuses it across layers;
- validates cache shape, dtype, device, and layer progress;
- adds CPU regression tests and a standalone legacy-vs-optimized microbenchmark that requires no model download.

The cache remains per request/session and grows on demand instead of reserving the maximum sequence length up front. No API, deploy configuration, model weight, or decoder dtype changes are included.

## Test Plan

```bash
pytest -q \
  tests/model_executor/models/audex/test_audex_causal_speech_decoder_cache.py \
  -m "core_model and cpu" \
  --run-level=core_model
```

Representative benchmark command:

```bash
python benchmarks/audex/bench_streaming_decoder_cache.py \
  --device cuda \
  --dtype float32 \
  --batch-size 1 \
  --hidden-dim 2048 \
  --depth 12 \
  --heads 32 \
  --total-frames 512 \
  --chunk-frames 5 \
  --warmups 2 \
  --iterations 10
```

The benchmark alternates legacy and optimized execution order between measured iterations. The checked-in legacy runner reproduces repeated K/V concatenation, per-layer mask construction, and per-layer RoPE position readback.

**vLLM Version:** `0.26.0`

**vLLM-Omni Commit:** `f64028769b43b91987c2d53e3f3a9e8fdfaedcb5`

Benchmark environment:

- 1x NVIDIA GeForce RTX 4090
- Driver 591.86
- PyTorch 2.11.0+cu130 / CUDA 13.0
- Audex decoder dimensions: hidden size 2048, 12 layers, 32 heads
- 2 warmups and 10 alternating measured iterations per configuration

### Sequence-length scaling

FP32, batch 1, 5 frames/chunk. Latencies are median ± population stddev per complete stream.

| Frames | Decoder calls | Legacy latency (ms) | Optimized latency (ms) | Speedup | Extra peak allocated (MiB) |
|---:|---:|---:|---:|---:|---:|
| 128 | 26 | 400.433 ± 15.598 | 246.925 ± 9.512 | 1.622x | 27.1 → 26.0 |
| 256 | 52 | 1019.739 ± 76.751 | 660.972 ± 86.546 | 1.543x | 54.5 → 52.0 |
| 512 | 103 | 1950.904 ± 150.964 | 1126.015 ± 155.918 | 1.733x | 108.9 → 104.0 |
| 1024 | 205 | 3613.092 ± 396.153 | 2321.342 ± 223.844 | 1.556x | 216.1 → 208.0 |
| 2048 | 410 | 6417.564 ± 68.153 | 4228.245 ± 15.446 | 1.518x | 432.1 → 416.0 |

### Chunk-size sensitivity

FP32, batch 1, 512 total frames.

| Frames/chunk | Decoder calls | Legacy latency (ms) | Optimized latency (ms) | Speedup | Extra peak allocated (MiB) |
|---:|---:|---:|---:|---:|---:|
| 1 | 512 | 8162.571 ± 729.178 | 4764.850 ± 730.387 | 1.713x | 108.5 → 104.0 |
| 2 | 256 | 3645.382 ± 262.222 | 2291.523 ± 11.625 | 1.591x | 108.7 → 104.0 |
| 5 | 103 | 1950.904 ± 150.964 | 1126.015 ± 155.918 | 1.733x | 108.9 → 104.0 |
| 10 | 52 | 1016.488 ± 95.669 | 655.039 ± 69.436 | 1.552x | 109.5 → 104.0 |
| 20 | 26 | 434.923 ± 34.968 | 268.875 ± 28.650 | 1.618x | 109.4 → 104.0 |
| 40 | 13 | 216.109 ± 20.641 | 144.148 ± 23.129 | 1.499x | 109.0 → 104.1 |

### Decoder tensor batch scaling

FP32, 512 total frames, 5 frames/chunk. This is decoder tensor batching, not online serving concurrency.

| Batch | Legacy latency (ms) | Optimized latency (ms) | Speedup | Extra peak allocated (MiB) |
|---:|---:|---:|---:|---:|
| 1 | 1950.904 ± 150.964 | 1126.015 ± 155.918 | 1.733x | 108.9 → 104.0 |
| 2 | 1574.488 ± 227.946 | 980.861 ± 140.427 | 1.605x | 216.1 → 208.0 |
| 4 | 1774.456 ± 56.734 | 1076.868 ± 18.714 | 1.648x | 432.3 → 416.0 |
| 8 | 1779.759 ± 22.294 | 1152.118 ± 20.528 | 1.545x | 865.8 → 832.0 |

### Dtype robustness

Batch 1, 512 total frames, 5 frames/chunk.

| Dtype | Legacy latency (ms) | Optimized latency (ms) | Speedup | Extra peak allocated (MiB) |
|---|---:|---:|---:|---:|
| FP32 | 1950.904 ± 150.964 | 1126.015 ± 155.918 | 1.733x | 108.9 → 104.0 |
| BF16 | 1358.388 ± 34.513 | 1046.986 ± 57.033 | 1.297x | 54.1 → 52.0 |

All 14 benchmark configurations reported maximum absolute output difference `0` between the legacy and optimized paths.

```text
7 passed in 0.41s
```

All diff-scoped pre-commit hooks pass, including Ruff, formatting, typos, pytest marker validation, and pickle-import validation.

### CODE REVIEW

**Comment by linyueqian** [comment]
@yuekaizhang ptal

File: vllm_omni/model_executor/models/audex/speech_decoder/modeling_audex_causal_speech_decoder.py [hunk 1]
```diff
@@ -49,19 +49,19 @@ def build_rope_cache(self, max_seq_len: int = 4096) -> None:
         cache = torch.stack([torch.cos(idx_theta), torch.sin(idx_theta)], dim=-1)
         self.register_buffer("cache", cache, persistent=False)
 
+    def prepare(self, needed_seq_len: int, device: torch.device) -> None:
+        """Ensure the RoPE cache is ready without reading a device tensor."""
+        if needed_seq_len < 0:
+            raise ValueError(f"needed_seq_len must be non-negative, got {needed_seq_len}")
+        if self.theta.device != device:
```
Comment (lishunyang12) [inline vllm_omni/model_executor/models/audex/speech_decoder/modeling_audex_causal_speech_decoder.py:56]: Rebuild RoPE in FP32 after dtype casts; BF16 cache growth currently merges positions.

Comment (pxljs) [reply vllm_omni/model_executor/models/audex/speech_decoder/modeling_audex_causal_speech_decoder.py]: thank u,fixed
